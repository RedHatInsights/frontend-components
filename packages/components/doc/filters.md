# Filters

## FilterInput
Implementation of the 'FilterInput' component for displaying filter inputs in toolbar filters. 

### Usage

Import FilterInput from this package. The `type` of input can be `radio` or `checkbox`. 

```JSX
import React from 'react';
import { FilterInput } from '@redhat-cloud-services/frontend-components';

class YourCmp extends React.Component {
  render() {
    return (
      <FilterInput
          aria-label="label"
          id="id"
          label="label"
          addRemoveFilters={ jest.fn() }
          param="param"
          type="radio"
          value="value"
          filters={ { param: 'value' } }
      />
    )
  }
}
```

## FilterDropdown
Implementation of the 'FilterDropdown' component for displaying filter dropdowns in toolbar filters. 

### Usage

Import FilterDropdown from this package.

```JSX
import React from 'react';
import { FilterDropdown } from '@redhat-cloud-services/frontend-components';

class YourCmp extends React.Component {
  render() {
    return (
      <FilterDropdown
          filters={ {} }
          addFilter={ console.log }
          removeFilter={ console.log }
          filterCategories={ [{ title: '', type: '', urlParam: '', values: [{ label: '', value: '' }]}] }
      />
    )
  }
}
```

