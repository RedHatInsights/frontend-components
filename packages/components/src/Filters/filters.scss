.ins-c-filter__dropdown .pf-c-dropdown__menu-item {
  &:hover {
    background-color: inherit !important;
    outline: inherit !important;
  }
  &:focus {
    background-color: inherit !important;
    outline: inherit !important;
  }
}