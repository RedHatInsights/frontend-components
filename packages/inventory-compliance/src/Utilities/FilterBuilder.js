import { conditionalFilterType } from '@redhat-cloud-services/frontend-components';
import {
    stringToId
} from './FilterConfigBuilder';

class FilterBuilder {
    constructor(filterConfig) {
        this.filterConfig = filterConfig;
        this.config = this.filterConfig.config;
    }

    buildFilterFilterString = (configItem, value) => {
        if (!value) { return []; }

        const { type, filterString } = configItem;

        switch (type) {
            case conditionalFilterType.text:
                return [ filterString(value) ];
            case conditionalFilterType.checkbox:
                return value.map((filter) => (filterString(filter)));
            default:
                return [];
        }
    }

    combineFilterStrings = (filterStringArray) => {
        const moreThanTwo = filterStringArray.map((f) => (f.length)).filter((fl) => (fl > 0)).length >= 2;
        return filterStringArray.map((fs) => (fs.join(' or '))).join(moreThanTwo ? ' and ' : '');
    }

    buildFilterString = (filters) => {
        const filterStringArray = this.config.map((configItem) => (
            this.buildFilterFilterString(configItem, filters[stringToId(configItem.label)])
        )).filter((f) => (f.length > 0));
        return this.combineFilterStrings(filterStringArray);
    }
}

export default FilterBuilder;
