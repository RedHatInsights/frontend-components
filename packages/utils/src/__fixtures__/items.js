export const severityLevels = ['high', 'medium', 'low', 'unknown'];

export default (count) => {
  let currentCount = 0;
  return [...new Array(count > 0 ? count : 1)]
    .map(() => {
      currentCount++;
      return {
        id: currentCount,
        severity: severityLevels[currentCount % 4],
        name: 'TEST ITEM #' + currentCount,
        description: 'DESCRIPTION',
      };
    })
    .sort();
};
