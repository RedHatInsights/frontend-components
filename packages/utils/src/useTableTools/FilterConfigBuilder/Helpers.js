export const stringToId = (string) => (
    string.split(' ').join('').toLowerCase()
);
