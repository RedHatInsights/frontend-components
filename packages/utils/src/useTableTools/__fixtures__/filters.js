export const exampleFilters = [{
    type: 'text',
    label: 'Name',
    filterString: (value) => (`name ~ ${value}`)
}, {
    type: 'checkbox',
    label: 'Compliant',
    filterString: (value) => (`compliant = ${value}`),
    items: [
        { label: 'Compliant', value: 'true' },
        { label: 'Non-compliant', value: 'false' }
    ]
}, {
    type: 'checkbox',
    label: 'Systems meeting compliance',
    filterString: (value) => {
        const scoreRange = value.split('-');
        return `compliance_score >= ${scoreRange[0]} and compliance_score <= ${scoreRange[1]}`;
    },
    items: [
        { label: '90 - 100%', value: '90-100' },
        { label: '70 - 89%', value: '70-89' },
        { label: '50 - 69%', value: '50-69' },
        { label: 'Less than 50%', value: '0-49' }
    ]
}];

export default [
    {
        type: 'text',
        label: 'Name',
        filter: (items, value) => (items.filter((item) => (
            item?.name.includes(value)
        )))
    },
    {
        type: 'checkbox',
        label: 'Chackbox Filter',
        items: [ 'OPTION 1', 'OPTION 2', 'OPTION 3' ].map((option) => ({
            label: option,
            value: option
        })),
        filter: (items) => (items)
    }
];
