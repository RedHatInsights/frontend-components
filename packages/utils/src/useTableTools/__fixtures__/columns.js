import { severityLevels } from './items';

export default [
    {
        title: 'Name',
        key: 'name',
        sortByProperty: 'name',
        renderFunc: (item) => (
            item.name
        )
    }, {
        title: 'Description',
        key: 'description'
    }, {
        title: 'Severity sorted by function',
        sortByFunction: (item) => (severityLevels.indexOf(item.severity)),
        renderFunc: (item) => (
            item.severity
        )
    }, {
        title: 'Severity by Array',
        sortByProperty: 'severity',
        sortByArray: severityLevels,
        renderFunc: (item) => (
            item.severity
        )
    }
];
