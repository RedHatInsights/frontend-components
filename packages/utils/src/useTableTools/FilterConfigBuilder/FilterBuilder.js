import { conditionalFilterType } from '@redhat-cloud-services/frontend-components/ConditionalFilter';
import { stringToId } from './Helpers';

class FilterBuilder {
    constructor(filterConfig) {
        this.filterConfig = filterConfig;
        this.config = this.filterConfig.config;
    }

    buildFilterFilterString = (configItem, value) => {
        const { type, filterString } = configItem;

        if (type !== 'hidden' && !value) { return []; }

        switch (type) {
            case conditionalFilterType.text:
                return [ filterString(value) ];
            case conditionalFilterType.checkbox:
                return value.map((filter) => (filterString(filter)));

            case conditionalFilterType.group:
                return filterString(value);

            case 'hidden':
                return filterString();

            default:
                return [];
        }
    }

    combineFilterStrings = (filterStringArray) => {
        const moreThanTwo = filterStringArray.map((f) => (f.length)).filter((fl) => (fl > 0)).length >= 2;
        return filterStringArray.map((fs) => (fs.join(' or '))).join(moreThanTwo ? ' and ' : '');
    }

    buildFilterString = (filters) => {
        const filterStringArray = this.config.map((configItem) => (
            this.buildFilterFilterString(configItem, filters[stringToId(configItem.label)])
        )).filter((f) => (f.length > 0));
        return this.combineFilterStrings(filterStringArray);
    }
}

export default FilterBuilder;
