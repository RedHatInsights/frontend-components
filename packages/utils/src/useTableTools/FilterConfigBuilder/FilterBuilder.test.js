import FilterConfigBuilder from './FilterConfigBuilder';
import FilterBuilder from './FilterBuilder';
import { exampleFilters } from '../__fixtures__/filters';

describe('buildFilterString', () => {
    const configBuilder = new FilterConfigBuilder(exampleFilters);
    let filterBuilder;

    beforeEach(() => {
        filterBuilder = new FilterBuilder(configBuilder);
    });

    it('returns a filterstring', () => {
        const exampleActiveFilters = {
            name: 'Name',
            compliant: [ true ],
            compliancescore: [ '0-49', '50-69' ]
        };

        expect(filterBuilder.buildFilterString(exampleActiveFilters)).toMatchSnapshot();
    });

    describe('filter building', () => {
        it('returns a base filter for name when searching', () => {
            const testExampleState = {
                name: 'Name'
            };
            expect(filterBuilder.buildFilterString(testExampleState)).toMatchSnapshot();
        });

        it('returns a filter for compliant', () => {
            const testExampleState = {
                compliant: [ true ]
            };
            expect(filterBuilder.buildFilterString(testExampleState)).toMatchSnapshot();
        });

        it('returns a filter for scores', () => {
            const testExampleState = {
                systemsmeetingcompliance: [ '0-49', '50-69' ]
            };
            expect(filterBuilder.buildFilterString(testExampleState)).toMatchSnapshot();
        });
    });
});
