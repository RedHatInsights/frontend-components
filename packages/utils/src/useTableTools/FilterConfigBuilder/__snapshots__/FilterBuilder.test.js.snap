// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildFilterString filter building returns a base filter for name when searching 1`] = `"name ~ Name"`;

exports[`buildFilterString filter building returns a filter for compliant 1`] = `"compliant = true"`;

exports[`buildFilterString filter building returns a filter for scores 1`] = `"compliance_score >= 0 and compliance_score <= 49 or compliance_score >= 50 and compliance_score <= 69"`;

exports[`buildFilterString returns a filterstring 1`] = `"name ~ Name and compliant = true"`;
